sdk/core/src/tokio/fs.rs

```diff
@@ -0,0 +1,165 @@
+use crate::{
+    request::Body,
+    seekable_stream::{SeekableStream, DEFAULT_BUFFER_SIZE},
+    setters,
+};
+use futures::{task::Poll, Future};
+use std::{cmp::min, io::SeekFrom, pin::Pin, sync::Arc, task::Context};
+use tokio::{
+    fs::File,
+    io::{AsyncReadExt, AsyncSeekExt, Take},
+    sync::Mutex,
+};
+
+#[derive(Debug)]
+pub struct FileStreamBuilder {
+    handle: File,
+    /// Offset into the file to start reading from
+    offset: Option<u64>,
+    /// Amount of data to read from the file
+    buffer_size: Option<usize>,
+    /// How much to buffer in memory during streaming reads
+    block_size: Option<u64>,
+}
+
+impl FileStreamBuilder {
+    pub fn new(handle: File) -> Self {
+        Self {
+            handle,
+            offset: None,
+            buffer_size: None,
+            block_size: None,
+        }
+    }
+
+    setters! {
+        // #[doc = "Offset into the file to start reading from"]
+        offset: u64 => Some(offset),
+        // #[doc = "Amount of data to read from the file"]
+        block_size: u64 => Some(block_size),
+        // #[doc = "Amount of data to buffer in memory during streaming reads"]
+        buffer_size: usize => Some(buffer_size),
+    }
+
+    pub async fn build(mut self) -> crate::Result<FileStream> {
+        let stream_size = self.handle.metadata().await?.len();
+
+        let buffer_size = self.buffer_size.unwrap_or(DEFAULT_BUFFER_SIZE);
+
+        let offset = if let Some(offset) = self.offset {
+            self.handle.seek(SeekFrom::Start(offset)).await?;
+            offset
+        } else {
+            0
+        };
+
+        let block_size = if let Some(block_size) = self.block_size {
+            block_size
+        } else {
+            stream_size - offset
+        };
+
+        let handle = Arc::new(Mutex::new(self.handle.take(block_size)));
+
+        Ok(FileStream {
+            handle,
+            buffer_size,
+            block_size,
+            stream_size,
+            offset,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+#[pin_project::pin_project]
+pub struct FileStream {
+    #[pin]
+    handle: Arc<Mutex<Take<File>>>,
+    pub stream_size: u64,
+    pub block_size: u64,
+    buffer_size: usize,
+    pub offset: u64,
+}
+
+impl FileStream {
+    async fn read(&mut self, slice: &mut [u8]) -> std::io::Result<usize> {
```

Comment:
What is the difference between this inherent method and `AsyncRead::read`? A comment explaining that here would probably be helpful.

Reply:
Added.